cpp/src/experimental/graph_view.cu

```diff
@@ -0,0 +1,316 @@
+               bool sorted_by_global_degree_within_vertex_partition,
+               bool do_expensive_check)
+  : detail::graph_base_t<vertex_t, edge_t, weight_t>(
+      handle, number_of_vertices, number_of_edges, is_symmetric, is_multigraph, is_weighted),
+    adj_matrix_partition_offsets_(adj_matrix_partition_offsets),
+    adj_matrix_partition_indices_(adj_matrix_partition_indices),
+    adj_matrix_partition_weights_(adj_matrix_partition_weights),
+    partition_(partition),
+    vertex_partition_segment_offsets_(vertex_partition_segment_offsets)
+{
+  // cheap error checks
+
+  auto comm_p_size     = this->get_handle_ptr()->get_comms().get_size();
+  auto comm_p_row_size = this->get_handle_ptr()->get_subcomm(comm_p_row_key).get_size();
+  auto comm_p_col_size = this->get_handle_ptr()->get_subcomm(comm_p_col_key).get_size();
+
+  CUGRAPH_EXPECTS(adj_matrix_partition_offsets.size() == adj_matrix_partition_indices.size(),
+                  "Invalid API parameter: adj_matrix_partition_offsets.size() and "
+                  "adj_matrix_partition_indices.size() should coincide.");
+  CUGRAPH_EXPECTS(
+    (is_weighted && (adj_matrix_partition_weights.size() == adj_matrix_partition_offsets.size())) ||
+      (!is_weighted && (adj_matrix_partition_weights.size() == 0)),
+    "Invalid API parameter: adj_matrix_partition_weights.size() should coincide with "
+    "adj_matrix_partition_offsets.size() (if is_weighted is true) or 0 (if is_weighted is false).");
+
+  CUGRAPH_EXPECTS(
+    (partition.hypergraph_partitioned &&
+     (adj_matrix_partition_offsets.size() == static_cast<size_t>(comm_p_row_size))) ||
+      (!(partition.hypergraph_partitioned) && (adj_matrix_partition_offsets.size() == 1)),
+    "Invalid API parameter: errneous adj_matrix_partition_offsets.size().");
+
+  CUGRAPH_EXPECTS((sorted_by_global_degree_within_vertex_partition &&
+                   (vertex_partition_segment_offsets.size() ==
+                    comm_p_col_size * (detail::num_segments_per_vertex_partition + 1))) ||
+                    (!sorted_by_global_degree_within_vertex_partition &&
+                     (vertex_partition_segment_offsets.size() == 0)),
+                  "Invalid API parameter: vertex_partition_segment_offsets.size() does not match "
+                  "with sorted_by_global_degree_within_vertex_partition.");
+
+  CUGRAPH_EXPECTS(partition.vertex_partition_offsets.size() == static_cast<size_t>(comm_p_size),
+                  "Invalid API parameter: erroneous partition.vertex_partition_offsets.size().");
+
+  // optinoal expensive checks
+
+  if (do_expensive_check) {
+    auto default_stream = this->get_handle_ptr()->get_stream();
+
+    auto comm_p_row_rank = this->get_handle_ptr()->get_subcomm(comm_p_row_key).get_rank();
+    auto comm_p_col_rank = this->get_handle_ptr()->get_subcomm(comm_p_col_key).get_rank();
+
+    edge_t number_of_local_edges_sum{};
+    for (size_t i = 0; i < adj_matrix_partition_offsets.size(); ++i) {
```

Comment:
Consider caching the `size()`. For a variety of (STL-like) containers, `size()` can be an expensive call.

Even better, if you _can_ `for_each() / transform()` with iterators on `adj_matrix_partition_offsets` that would lead to more readable code (e.g., by refactoring the checkers into separate functors that can be sequentially called on each iterator instance)

Reply:
adj_matrix_partition_offsets is std::vector; so size() is cheap. std::vector has three variables: pointer to the memory, size & capacity. So size() is just a getter. size() can be expensive for std::list kinda data structures but it's not a case here.

I tried to refactor this to two separate std::any_of calls and one transform_reduce call, but the problem is that I need to execute
```
raft::update_host(&number_of_local_edges,
                        adj_matrix_partition_offsets[i] + (major_last - major_first),
                        1,
                        default_stream);
```
2x more. This adds performance overhead and repeating this code 2x is also not desirable.

I can use thrust::make_zip_iterator (standard C++ does not support zip_iterator and I need to use boost instead, and I want to avoid this) and call thrust function with thrust::host execution policy, but I sort of want to avoid this; I guess it is common in RAPIDS to use std::algorithm for accessing host data and thrust::algorithm for accessing device data, and I did not want to go against this pattern. Let me know what you think.